iroh-sync/src/ranger.rs

```diff
@@ -0,0 +1,1322 @@
+//! Implementation of Set Reconcilliation based on
+//! "Range-Based Set Reconciliation" by Aljoscha Meyer.
+//!
+
+use std::collections::BTreeMap;
+use std::fmt::Debug;
+use std::marker::PhantomData;
+use std::{cmp::Ordering, convert::Infallible};
+
+use serde::{Deserialize, Serialize};
+
+/// Stores a range.
+///
+/// There are three possibilities
+/// - x, x: All elements in a set, denoted with
+/// - [x, y): x < y: Includes x, but not y
+/// - S \ [y, x) y < x: Includes x, but not y.
+/// This means that ranges are "wrap around" conceptually.
+#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize, Default)]
+pub struct Range<K> {
+    x: K,
+    y: K,
+}
+
+impl<K> Range<K> {
+    pub fn x(&self) -> &K {
+        &self.x
+    }
+
+    pub fn y(&self) -> &K {
+        &self.y
+    }
+
+    pub fn new(x: K, y: K) -> Self {
+        Range { x, y }
+    }
+
+    pub fn map<X>(self, f: impl FnOnce(K, K) -> (X, X)) -> Range<X> {
+        let (x, y) = f(self.x, self.y);
+        Range { x, y }
+    }
+}
+
+impl<K> From<(K, K)> for Range<K> {
+    fn from((x, y): (K, K)) -> Self {
+        Range { x, y }
+    }
+}
+
+pub trait RangeKey: Sized + Ord + Debug {
+    /// Is this key inside the range?
+    fn contains(&self, range: &Range<Self>) -> bool {
+        contains(self, range)
+    }
+}
+
+/// Default implementation of `contains` for `Ord` types.
+pub fn contains<T: Ord>(t: &T, range: &Range<T>) -> bool {
```

Comment:
why is this available both directly and via the trait?

Reply:
Because they take different arguments, and the standalone version is useful to implement the trait on other types.